subprojects/core-api/src/main/java/org/gradle/vcs/VcsMapping.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.vcs;
+
+import org.gradle.api.Incubating;
+import org.gradle.api.artifacts.component.ComponentSelector;
+
+/**
+ * A dependency mapping provided by a VCS repository.
+ *
+ * @since 4.3
+ */
+@Incubating
+public interface VcsMapping {
```

Comment:
This has drifted a little bit from the original design. I'm not sure now how a build master would provide an implementation for VcsMapping that would implement some pattern her company uses for mapping from a component selector to a vcs repository. Let's discuss.

Reply:
They wouldn't provide an implementation.  We have the `VcsMappings` DSL to configure a `VcsMapping` instance we create to map a `ComponentSelector` to a `VersionControlSpec`.

Reply:
Yeah, but what if the company has a predictable pattern of how to map from a `ComponentSelector` to a `VersionControlSpec` for their internally-hosted GitHub Enterprise instance. They shouldn't have to add a `VcsMapping` for each repository. They should be able to provide a `VcsMapping` implementation which, given a `ComponentSelector` can return the appropriate `VersionControlSpec`